src/libclipper/include/clipper/future.hpp

```diff
@@ -0,0 +1,150 @@
+#ifndef CLIPPER_LIB_FUTURE_HPP
+#define CLIPPER_LIB_FUTURE_HPP
+
+#include <atomic>
+#include <cassert>
+#include <memory>
+#include <thread>
+#include <vector>
+
+#include "boost/thread.hpp"
+
+namespace clipper {
+
+namespace future {
+
+/*
+ * The first element in the pair is a future that will complete when
+ * all the futures provided have completed. The second element is a vector
+ * of futures that have the same values and will complete at the same time
+ * as the futures passed in as argument.
+ *
+ * \param futures A list of futures to compose
+ * \param num_completed A counter indicating the number of futures that
+ * have completed so far. The reason this is an argument to the function
+ * instead of internal state is that the caller of `when_all` must make sure
+ * that the counter stays in scope for the lifetime of all the futures.
+*/
+template <class T>
+std::pair<boost::future<void>, std::vector<boost::future<T>>> when_all(
+    std::vector<boost::future<T>> futures,
+    std::shared_ptr<std::atomic<int>> num_completed) {
+  if (futures.size() == 0) {
+    return std::make_pair(boost::make_ready_future(),
+                          std::vector<boost::future<T>>{});
+  }
+  int num_futures = futures.size();
+  auto completion_promise = std::make_shared<boost::promise<void>>();
+  std::vector<boost::future<T>> wrapped_futures;
+  for (auto f = futures.begin(); f != futures.end(); ++f) {
+    wrapped_futures.push_back(f->then(
+        [num_futures, completion_promise, num_completed](auto result) mutable {
```

Comment:
Yeah you're right, there is a race condition. Good catch!

(Documenting here to be explicit)

Consider the case where there are 2 futures (so `num_futures = 2`) and neither has completed yet.

They may both be completed simultaneously in separate threads. They will both fail the test in the if statement because `num_completed + 1 < num_futures` (0 + 2 < 2). Then _after the test_ they will both atomically update `num_completed`. The increment to `num_completed` will be safe because it's an atomic add, but now both futures have completed so the future composition should be marked as finished, and yet the `completion_promise` will never be set.

The fix is simple:
replace the if-else statement with:

```
if (num_completed.fetch_add(1) + 1== num_futures) {
    completion_promise->set_value();
}
```

Consider the previous example with two futures. Because `num_completed` is atomic, the two `fetch_add` calls will be serial. One of them is guaranteed to be the second call, in which case `num_completed.fetch_add(1)` returns 1 (the previous value of the atomic variable), so the conditional will always evaluate to true exactly once.